### ISSUE TITLE
`exp show`: Distinguish between different dependencies

### ISSUE THREAD

**Issue by dberenbaum** [iterative/dvc#7575]
@daavoo @dberenbaum 

Quick ask - in Studio we can distinguish DVC-tracked and intermediate artifacts (to show only them by default), can we add those signals here too please?

_Originally posted by @shcheklein in https://github.com/iterative/vscode-dvc/issues/1183#issuecomment-1062436116_

**Comment by dberenbaum** [comment]
@shcheklein @mattseddon I'm not clear on how Studio decides what files to show. From what I can tell, it's anything with a `.dvc` file plus outputs of any pipeline stages. 

@daavoo AFAIK the experiments table instead shows dependencies of any pipeline stages.

Do we need to reconcile these and choose one or the other? Do we need options in any of dvc, studio, or vs code to decide between them?

**Comment by daavoo** [comment]
I thought the request was to differentiate between dependencies tracked by git vs. dependencies tracked by DVC (the former are not shown in Studio table)

**Comment by daavoo** [comment]
## Differences

- `src/*`

Missing in Studio - Present in DVC

These are files explicitly marked as DVC dependencies in `dvc.yaml` but are not shown in Studio (because they are tracked by git?). 

- `prc.json` / `roc.json` / `scores.json`

Present in Studio - Missing in DVC

These are `metrics` and `plots` outputs (tracked by git in this case).

---

## Exp show behavior

`dvc exp show --json` returns `deps` and `outs` separately but I decided to only show `deps` in the initial implementation because that's what I considered valuable information when comparing experiments (also to prevent noise).

IMO, this makes sense semantically because users explicitly understand and define the concepts of `deps` and `outs` in `dvc.yaml`. The docs explicitly mention `dependencies` as part of the table https://dvc.org/doc/command-reference/exp/show#description

---

## Studio behavior

I'm actually confused by Studio logic (hidden from the user):

- There is no concept or way to differentiate  `deps` or `outs` 
- Doesn't show git-tracked dependencies but does show git-tracked outputs
- Show `metrics` and `plots` as `Files`

---

## 1) DVC-tracked files

The only `dvc-tracked-file` that is currently present in Studio but would be missing in `exp show` would be one that it's not connected to the pipeline at all. 

- [ ] If it's not connected to the pipeline, I'm not sure why it would make sense to be present in the `exp show` output, but I can include them in `--json` output.

- [ ] Including `metrics` and `plots` files as outputs don't make sense to me, but I can include them in the `--json` output.

- [ ] If VSCode wants to filter out git-tracked dependencies, I can include this info in the `--json` output. 

However, I personally think that the current `exp show` is more useful for comparing experiments than the current Studio `Files`.

---

## 2) intermediate artifacts (stage outputs)

Stage outputs are already present in `--json` `outs` section . I don't think what additional signal would be needed.

**Comment by daavoo** [comment]
@dberenbaum @shcheklein @mattseddon Can I get confirmation that you want the 3 checkboxes to be implemented and that you don't need additional info for `2) intermediate artifacts (stage outputs)`?

**Comment by mattseddon** [comment]
@daavoo we'll discuss tomorrow at the cross-team meeting and come back to you, thanks.

**Comment by dberenbaum** [comment]
> If it's not connected to the pipeline, I'm not sure why it would make sense to be present in the exp show output, but I can include them in the outs of --json output.
>
> Including metrics and plots files as outputs don't make sense to me, but I can include them in the outs of --json output.

I'm also unclear on whether these should be included. Do you think the VS Code experiments table would ever need either of these for the experiments table @shcheklein @mattseddon?

My worry is that adding them to `--json` is basically adding junk from the viewpoint of a CLI user in that it's not really helpful experiment info and it doesn't show up in the CLI table or other output formats.

> If VSCode wants to filter out git-tracked dependencies, I can include this info in the deps of --json output.

👍 

> @dberenbaum @shcheklein @mattseddon Can I get confirmation that you want the 3 checkboxes to be implemented and that you don't need additional info for `2) intermediate artifacts (stage outputs)`?

@shcheklein You mentioned that Studio filters out intermediate artifacts, so it seems like this is useful info, right?

I'm still a bit unclear as to how to define an intermediate artifact. In example-get-started, it looks like `data/prepared` is not an intermediate artifact because it's an output of the first stage in the pipeline, but `data/features` is an intermediate artifact because it's an output of an intermediate stage. Is that right?

**Comment by dberenbaum** [comment]
> In example-get-started, it looks like `data/prepared` is not an intermediate artifact because it's an output of the first stage in the pipeline

Clarified with @shcheklein that this is a bug in Studio, and `data/prepared` should be considered an intermediate artifact. So AFAIK intermediate artifacts are those defined as either dependencies or outputs in any intermediate stage (stages which are both upstream and downstream of other stages).

**Comment by daavoo** [comment]
> If it's not connected to the pipeline, I'm not sure why it would make sense to be present in the exp show output, but I can include them in the outs of --json output.

Files tracked with `dvc add` or `dvc import` are already present in the `outs` section even if they are not connected to the pipeline, we just don't show `outs` at all in the table.

**Comment by daavoo** [comment]
I added in https://github.com/iterative/dvc/pull/7690 what I believe is all the info required by VSCode to make decisions. 

I extended the `--json` output with additional info (`fs_path`, `use_cache` for outputs) that was already being collected. I think we should not further complicate the collection of data that happens in `repo.experiments.show`.

About the original 2 requests, here is how I would implement the logic in Python, which I believe could be easily translated and implemented on VSCode after parsing the `--json` output:

```python
for dep in res[“deps”].values():
    for out in res[“outs”].values():
        if dep[“fs_path”] == out[“fs_path”]:
            # THIS IS AN INTERMEDIATE ARTIFACT
            if out[“use_cache”]:
                # DEP IS DVC TRACKED
            else:
                # DEP IS GIT-TRACKED 
```

An alternative would be to add this snippet to the DVC codebase:

```python
for dep in res[“deps”].values():
    for out in res[“outs”].values():
        if dep[“fs_path”] == out[“fs_path”]:
            # THIS IS AN INTERMEDIATE ARTIFACT
            dep["intermediate"] = True
            out["intermediate"] = True
            if out[“use_cache”]:
                # DEP IS DVC TRACKED
                dep["use_cache"] = True
            else:
                # DEP IS GIT-TRACKED 
                dep["use_cache"] = False
```

Given that the rules about deciding what files should be included appear to be currently unclear and product-dependent I would prefer not to include this logic in DVC.

**Comment by pmrowla** [comment]
@daavoo Looking at the existing code, there's a bug where we don't handle the case where multiple deps/outs can have the same `def_path` (which is a path relative to a stage's `wdir`).

Basically, given the following pipeline:
```yaml
stages:
  foo:
    wdir: foo
    outs:
      - data.csv
  bar:
    wdir: bar
    outs:
      - data.csv
```

The two separate outputs `foo/data.csv` and `bar/data.csv` will both have `def_path = data.csv`. The same situation can occur without using `wdir` if you have multiple pipelines in the same DVC repo, and then have outputs named `data.csv` in each pipeline (since the real fs paths will be relative to each separate `dvc.yaml`, but `def_path` will always be `data.csv`).

This currently results in only one of the outputs being included in the `exp show` data:

```json
{
  "5ff68336bb7e6feb12a08476a0252e4ac439bf6d": {
    "baseline": {
      "data": {
        "timestamp": "2022-05-23T15:15:46",
        "deps": {},
        "outs": {
          "data.csv": {
            "hash": "c157a79031e1c40f85931829bc5fc552",
            "size": 4,
            "nfiles": null
          }
        },
        "queued": false,
        "running": false,
        "executor": null,
        "metrics": {},
        "name": "main"
      }
    }
  }
}
```

The `fs_path` change does not resolve this issue, since the `outs` entry is a dictionary keyed on `def_path`.

**Comment by pmrowla** [comment]
IMO this stems from the underlying problem of trying to shove everything about a DVC repo into flat table rows for UI purposes, and what we should really be doing is outputting a json schema that actually reflects the repo/pipeline structure.

Right now, each experiment/commit has a top level `outs` and `deps` entry that  just contains the flat dict (that should probably be a list) of all the files in the experiment. This makes sense when it is being consumed by the DVC CLI, since we are restricted by the fact that it is CLI, and at some point we do have to flatten everything so that it can be rendered in a flat CLI table.

However, when we are outputting information meant to be consumed programmatically by other products (whether its vscode/studio or something else entirely), I think it would make more sense for the experiment/commit entries to look more like `dvc.lock`, where everything is grouped into dicts by pipelines and then stages, and then each stage has its own lists of `deps`/`outs`.

This all goes in line with my thinking that ideally, vscode should not be consuming `exp show --json`. What we really want is a command that outputs the raw experiment data, and not in a way that is pre-formatted/flattened for DVC CLI purposes. So basically a command that outputs a json version of what goes into `dvc.lock`s (i.e. the DVC repo index for a given experiment). And then the vscode extension should be formatting/flattening/etc the experiment data however they need, in a way that makes sense based on how they want render it in their GUI.

**Comment by pmrowla** [comment]
IMO this issue is also applicable to metrics/params, where we flatten everything into top level `metrics`/`params` entries as much as possible (since again, that's how we display it in the CLI), but with regard to other products, I think we should be exposing the raw data the way it is actually stored in the DVC repo (as a part of the index), and then letting the consuming product decide how to flatten/render it.

see also: https://github.com/iterative/dvc/pull/7613 for a similar discussion about how (and whether or not) metrics/params/repo index should be exposed via python APIs

**Comment by daavoo** [comment]
>  basically a command that outputs a json version of what goes into dvc.locks (i.e. the DVC repo index for a given experiment). 

Expand `dvc stage list` to have `--json` and return this? 

> And then the vscode extension should be formatting/flattening/etc the experiment data however they need

This would require vscode to call the command with all experiment ids as targets, right?. 

So, they would still need to retrieve all experiment ids from a command and then pass those ids to the "index json" command

**Comment by dberenbaum** [comment]
Agreed @pmrowla but is there a short-term fix we can implement? Should we open a separate issue/discussion for long-term needs of how VS Code should collect experiment info from DVC?

**Comment by pmrowla** [comment]
In the short term, @daavoo's changes to add more fields to the `exp show --json` output should work for now.

But at some point we need to move beyond implementing more and more workarounds in `exp show --json`. All of these short term fixes still require engineering hours spent on both the DVC and vscode side for a "solution" that should really be removed and replaced entirely.

**Comment by mattseddon** [comment]
I agree with @pmrowla. Can we plan to all get together after the VS Code release and look at how best to make this work?
